lib/rotoscope/call_logger.rb

```diff
@@ -88,34 +102,59 @@ def state
     private
 
     def log_call(call)
-      caller_path = call.caller_path || ''
-      return if blacklist.match?(caller_path)
+      return if blacklist.match?(filepath(call))
       return if self == call.receiver
 
-      caller_class_name = call.caller_class_name || '<UNKNOWN>'
+      buffer = @output_buffer
+      buffer.clear
+
+      if @header.size > 1
+        @header[0...-1].each do |head|
+          buffer << '"' << send(head, call) << '",'
```

Comment:
This `send` calls will be slower than a normal method call due to not being able to leverage the inline method cache.  We could benchmark to see if it will be a problem though, since I'm not sure how significant the impact will be.

To avoid that problem, we could eval a block specifically for the given headers in the initializer, since this is such a hot path.  E.g. you could do this following in the initializer

```ruby
append_lines = headers.map do |header|
  %{<< '"' << #{header}(call) << '"'}
end
@append_headers_block = eval <<~RUBY
  lambda do |buffer, call|
    buffer << #{append_lines.join(" ',' ")} << "\\n"
  end
RUBY
```

and then do this here

```ruby
@append_headers_block.call(buffer, call)
io.write(buffer)
```

That would allow all the calls to leverage the ruby's inline method cache.

It would still introduce some additional calls, which don't seem like they will be as much of a problem, but could be avoided by inlining the ruby code to append each header instead of a call to a method that does that.  That should make the performance of the code similar to what it is now, aside from the call to the block eval generated block.

Reply:
So I've done some benchmarking.
master
```
       user     system      total        real
  20.444025   1.288119  21.732144 ( 21.897433)
```

current state

```
       user     system      total        real
  30.478693   1.361850  31.840543 ( 31.981835)
```

your suggestion
```
       user     system      total        real
  27.358513   1.526877  28.885390 ( 29.151419)
```

switch statement
```
```

The benchmarking code looked as follows:
```
    Benchmark.bm do |b|
      b.report do
        rotoscope_trace do
          1000000.times do
            e = Example.new
            e.normal_method
            e.yielding_method { e.normal_method }
            Example.singleton_method
          end
        end
      end
    end
```

It is considerably slower, so I think I'm going to abandon this PR and try to make this work in our script by monkey patching rotoscope.

Reply:
There should be no need to monkey patch rotoscope.  You can just not use `Rotoscope::CallLogger`, since it is just ruby code wrapping the `Rotoscope` class.  You could just use `Rotoscope::CallLogger` as a template for your custom call logger, then you can rip out everything you don't need and add in any custom code (e.g. filtering) that you want.  That is what we did for our component instrumentation in Shopify core.